### PR TITLE
Wait for required services instead of aborting

### LINKED ISSUE #81: Modbus Timeout kills launch process
### Commit
 05d7a01 

### Steps to reproduce
Start launch file on low-end pc with a large robot model. If the startup process takes more than 5 seconds, the modbus node dies.

### Expected behavior
Node waits until ROS is ready. There should not be a timeout at all, but just a warning to notify the user: "Service xxx not yet available, waiting...".

### Observed behavior
Modbus node dies and takes down ROS.

### Logfiles

```
[ INFO] [1552301651.010005889]: waitForService: Service [/prbt/manipulator_joint_trajectory_controller/hold] has not been advertised, waiting...
process[usb_cam-11]: started with pid [4544]
process[aruco_detect-12]: started with pid [4551]
process[ref_fid_publisher-13]: started with pid [4553]
process[fiducial_tf_converter-14]: started with pid [4567]
[ INFO] [1552301651.281068524]: Loading robot model 'prbt'...
[ WARN] [1552301651.281614825]: Skipping virtual joint 'FixedBase' because its child frame 'prbt_base_link' does not match the URDF frame 'base'
[ INFO] [1552301651.281992211]: No root/virtual joint specified in SRDF. Assuming fixed joint
[ INFO] [1552301651.980708962]: rviz version 1.12.16
[ INFO] [1552301651.980808438]: compiled against Qt version 5.5.1
[ INFO] [1552301651.980838067]: compiled against OGRE version 1.9.0 (Ghadamon)
[ INFO] [1552301652.355542478]: Using fixed control period: 0.010000000
[ INFO] [1552301652.587717879]: Initializing XXX
[ INFO] [1552301652.588118552]: Current state: 1 device error: system:0 internal_error: 0 (OK)
[ INFO] [1552301652.588401682]: Current state: 2 device error: system:0 internal_error: 0 (OK)
EMCY: 83#0090212500000044
EMCY: 84#0090212500000044
EMCY: 85#0090212500000044
[ INFO] [1552301652.812759987]: Stereo is NOT SUPPORTED
[ INFO] [1552301652.813402492]: OpenGl version: 3 (GLSL 1.3).
EMCY: 86#0090212500000044
EMCY: 87#0090212500000044
EMCY: 88#0090212500000044
ER: 33
[ INFO] [1552301653.775363515]: Current state: 2 device error: system:125 internal_error: 0 (OK)
[ INFO] [1552301653.775458465]: Current state: 1 device error: system:125 internal_error: 0 (OK)
[ INFO] [1552301653.775812399]: Current state: 1 device error: system:0 internal_error: 0 (OK)
[ INFO] [1552301653.775892800]: Current state: 0 device error: system:0 internal_error: 0 (OK)
[ INFO] [1552301653.775944583]: Current state: 0 device error: system:0 internal_error: 0 (OK)
[robot_init-6] process has finished cleanly
log file: /home/jgruner/.ros/log/f8ab4628-43eb-11e9-a364-8439be69721c/robot_init-6*.log
[ INFO] [1552301654.292010270]: Publishing maintained planning scene on 'monitored_planning_scene'
[ INFO] [1552301654.310662849]: MoveGroup debug mode is OFF
Starting context monitors...
[ INFO] [1552301654.310756724]: Starting scene monitor
[ INFO] [1552301654.333239512]: Listening to '/planning_scene'
[ INFO] [1552301654.333345179]: Starting world geometry monitor
[ INFO] [1552301654.356504353]: Listening to '/collision_object' using message notifier with target frame '/base '
[ INFO] [1552301654.373932880]: Listening to '/planning_scene_world' for planning scene world geometry
[ INFO] [1552301654.976778466]: Listening to '/attached_collision_object' for attached collision objects
Context monitors started.
[ INFO] [1552301654.993191516]: Reading limits from namespace /robot_description_planning
[ INFO] [1552301655.173408541]: Available plugins: pilz::PlanningContextLoaderCIRC pilz::PlanningContextLoaderLIN pilz::PlanningContextLoaderPTP 
[ INFO] [1552301655.173500376]: About to load: pilz::PlanningContextLoaderCIRC
[ INFO] [1552301655.177550915]: Registered Algorithm [CIRC]
[ INFO] [1552301655.178015249]: About to load: pilz::PlanningContextLoaderLIN
[ INFO] [1552301655.190567122]: Registered Algorithm [LIN]
[ INFO] [1552301655.190657450]: About to load: pilz::PlanningContextLoaderPTP
[ INFO] [1552301655.199674187]: Registered Algorithm [PTP]
[ INFO] [1552301655.200266448]: Using planning interface 'Simple Command Planner'
terminate called after throwing an instance of 'prbt_hardware_support::PilzStoModbusAdapterNodeException'
  what():  Controller service manipulator_joint_trajectory_controller/hold not available. STO adapter node can not function properly. Shutting down ROS ...
================================================================================REQUIRED process [prbt/sto_modbus_adapter_node-8] has died!
process has died [pid 4510, exit code -6, cmd /opt/ros/kinetic/lib/prbt_hardware_support/sto_modbus_adapter_node __name:=sto_modbus_adapter_node __log:=/home/jgruner/.ros/log/f8ab4628-43eb-11e9-a364-8439be69721c/prbt-sto_modbus_adapter_node-8.log].
log file: /home/jgruner/.ros/log/f8ab4628-43eb-11e9-a364-8439be69721c/prbt-sto_modbus_adapter_node-8*.log
Initiating shutdown!
================================================================================
```

### PR DESCRIPTION
**Pull request by jschleicher**
See example for *_DELAYED_THROTTLE in
http://wiki.ros.org/roscpp/Overview/Logging

Fixes #81

Please especially review the test cases I've changed: Instead of `EXPECT_THROW` they now wait for the service in seperate thread.

### CODE REVIEW

File: prbt_hardware_support/CMakeLists.txt [hunk 1]
```diff
@@ -206,7 +206,7 @@ if(CATKIN_ENABLE_TESTING)
   if(ENABLE_COVERAGE_TESTING)
     include(CodeCoverage)
     APPEND_COVERAGE_COMPILER_FLAGS()
-    set(COVERAGE_EXCLUDES "*/${PROJECT_NAME}/test*" "*/devel/include/${PROJECT_NAME}/*")
+    set(COVERAGE_EXCLUDES "*/${PROJECT_NAME}/test*" "*/ModbusMsgInStamped.h")
```
Comment (martiniil) [inline prbt_hardware_support/CMakeLists.txt:209]: Please pose a separate PR for this.

File: prbt_hardware_support/src/sto_modbus_adapter.cpp [hunk 2]
```diff
@@ -35,37 +35,36 @@ const std::string PilzStoModbusAdapterNode::HALT_SERVICE = "driver/halt";
...
 {
-  // Attach to hold service
-  if (!ros::service::waitForService(HOLD_SERVICE, ros::Duration(WAIT_FOR_SERVICE_TIMEOUT_S)))
+  // Wait for services
+  while (!ros::service::waitForService(HALT_SERVICE, ros::Duration(RETRY_SERVICE_CONNECTION_TIME_S)) && ros::ok())
   {
-    throw PilzStoModbusAdapterNodeException("Controller service " + HOLD_SERVICE
-                     + " not available. STO adapter node can not function properly. Shutting down ROS ...");
+  ROS_WARN_STREAM_DELAYED_THROTTLE(WAIT_FOR_SERVICE_TIMEOUT_S, "Waiting for driver service " + HALT_SERVICE
+                                   + " to become available... Please check, if the motor driver is up and running.");
```
Comment (martiniil) [inline prbt_hardware_support/src/sto_modbus_adapter.cpp]: Please format the code (wrong indent)

File: prbt_hardware_support/src/sto_modbus_adapter.cpp [hunk 3]
```diff
@@ -35,37 +35,36 @@ const std::string PilzStoModbusAdapterNode::HALT_SERVICE = "driver/halt";
...
 
-  // Attach to unhold service
   if(!ros::service::waitForService(UNHOLD_SERVICE, ros::Duration(WAIT_FOR_SERVICE_TIMEOUT_S)))
   {
-    ROS_ERROR_STREAM("Controller service " << UNHOLD_SERVICE
-                     << " not available. STO adapter node can not function properly. You can't unhold the controller"
-                        " once it went to hold mode.");
+  ROS_WARN_STREAM("Controller service " << UNHOLD_SERVICE
+                   << " not available. STO adapter node can not function properly. You can't unhold the controller"
+                      " once it went to hold mode.");
```
Comment (martiniil) [inline prbt_hardware_support/src/sto_modbus_adapter.cpp]: Fix indent.

File: prbt_hardware_support/src/sto_modbus_adapter.cpp [hunk 4]
```diff
@@ -35,37 +35,36 @@ const std::string PilzStoModbusAdapterNode::HALT_SERVICE = "driver/halt";
...
   }
-  unhold_srv_client_ = nh.serviceClient<std_srvs::Trigger>(UNHOLD_SERVICE);
-
-  // Attach to recover service
   if(!ros::service::waitForService(RECOVER_SERVICE, ros::Duration(WAIT_FOR_SERVICE_TIMEOUT_S)))
   {
-    ROS_ERROR_STREAM("Driver service " << RECOVER_SERVICE
-                     << " not available. Driver won't auto-recover after STO clearance.");
+  ROS_WARN_STREAM("Driver service " << RECOVER_SERVICE
+                   << " not available. Driver won't auto-recover after STO clearance.");
```
Comment (martiniil) [inline prbt_hardware_support/src/sto_modbus_adapter.cpp]: Fix indent.

File: prbt_hardware_support/test/unittests/unittest_sto_modbus_adapter.cpp [hunk 5]
```diff
@@ -131,17 +141,22 @@ TEST_F(PilzStoModbusAdapterTest, testSetup)
...
  */
 TEST_F(PilzStoModbusAdapterTest, testSetupNoDisableService)
 {
   manipulator_.advertiseHoldService(nh_, HOLD_SERVICE_T);
   manipulator_.advertiseUnholdService(nh_, UNHOLD_SERVICE_T);
   manipulator_.advertiseRecoverService(nh_, RECOVER_SERVICE_T);
 
-  EXPECT_THROW(PilzStoModbusAdapterNode adapter_node(nh_), PilzStoModbusAdapterNodeException);
+  auto t = asyncConstructor();
+  ros::Duration(5.5).sleep();
```
Comment (martiniil) [inline prbt_hardware_support/test/unittests/unittest_sto_modbus_adapter.cpp]: Maybe add a small comment explaining the choice of the duration.

**Review by martiniil — CHANGES_REQUESTED** [review]
I approve the changes. There are some formatting issues only. I think the description of the following test cases should be updated:
https://github.com/PilzDE/pilz_robots/blob/757de3ed7a66a0e6cd2dcf52c6ddc20e47848202/prbt_hardware_support/test/unittests/unittest_sto_modbus_adapter.cpp#L163
https://github.com/PilzDE/pilz_robots/blob/757de3ed7a66a0e6cd2dcf52c6ddc20e47848202/prbt_hardware_support/test/unittests/unittest_sto_modbus_adapter.cpp#L177

File: prbt_hardware_support/test/unittests/unittest_sto_modbus_adapter.cpp (hunk 5 above)
Comment (jschleicher) [reply prbt_hardware_support/test/unittests/unittest_sto_modbus_adapter.cpp]: Yep, good point.

File: prbt_hardware_support/CMakeLists.txt (hunk 1 above)
Comment (jschleicher) [reply prbt_hardware_support/CMakeLists.txt:209]: This line reverts dcc5ae1d1.
I had opened #121 as separate PR to fix the exclude by not installing test headers. Rebasing onto current melodic devel brought your change from dcc5ae1d1 as revert in here. But if you set great value on it, I'll open another one-line PR, just let me know!

**Comment by jschleicher** [comment]
Please squash-merge the additional commits.

File: prbt_hardware_support/CMakeLists.txt (hunk 1 above)
Comment (martiniil) [reply prbt_hardware_support/CMakeLists.txt:209]: I see

**Comment by jschleicher** [comment]
@martiniil ping

**Review by martiniil — APPROVED** [review]

**Pull request merged**